PowerShellTools.HostService/ServiceManagement/Debugging/PowerShellDebuggingService.cs

```diff
@@ -204,80 +241,178 @@ public void AttachToRunspace(uint pid)
+                    for (int i = 0; i < result.Count; i += 2)
+                    {
+                        dynamic pid = result.ElementAt(i);
+                        dynamic name = result.ElementAt(i + 1);
+                        validProcesses.Add(new KeyValuePair<uint, string>(uint.Parse(pid.ToString()), name.ToString()));
+                    }
+
+                    // Exit the remote session and return results back to RemoteEnumDebugProcess
+                    _currentPowerShell.Commands.Clear();
+                    _currentPowerShell.AddScript(string.Format(DebugEngineConstants.ExitRemoteSessionDefaultCommand));
+                    _currentPowerShell.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    ServiceCommon.Log("Error connecting to remote machine; " + ex.ToString());
+                    return null;
+                }
+            }
+            return validProcesses;
         }
 
         /// <summary>
-        /// Detaches the HostService from a local runspace
+        /// Attaches the HostService to a remote runspace already in execution
         /// </summary>
-        public void DetachFromRunspace()
+        public void AttachToRemoteRunspace(uint pid, string remoteName)
         {
             if (_callback == null)
             {
                 _callback = OperationContext.Current.GetCallbackChannel<IDebugEngineCallback>();
             }
 
-            try
+            using (_currentPowerShell = PowerShell.Create())
             {
-                // attempt to gracefully detach the deugger
-                ClearBreakpoints();
-                ExecuteDebuggingCommand("detach", false);
+                try
+                {
+                    // enter into a remote session
+                    _currentPowerShell.Runspace = _runspace;
+                    _currentPowerShell.AddScript(string.Format(DebugEngineConstants.EnterRemoteSessionDefaultCommand, remoteName));
+                    _currentPowerShell.Invoke();
+                }
+                catch (Exception e)
+                {
+                    ServiceCommon.Log("Unable to connect to the remote machine; " + e.Message);
+                    return;
+                }
             }
-            catch (Exception e)
+
+            // now that we are in the remote session we can attach to the runspace
+            AttachToRunspace(pid);
+        }
+
+        /// <summary>
+        /// Detaches the HostService from a remote runspace
+        /// </summary>
+        public bool DetachFromRemoteRunspace()
+        {
+            // detach from the runspace
+            if (!DetachFromRunspace())
             {
-                // if program is running we must use stop to force the debugger to detach
-                ServiceCommon.Log("Script currently in execution, must use stop to end debugger");
-                _currentPowerShell.Stop();
+                return false;

```

Comment:
In general, VS will prompt the user asking if they would like VS to attempt to force end debugging. I can get a more comprehensive list of experiences (depending on where the DetachFromRunspace() returns false) in a bit though.